File: parking-lot/go-without-multithreading/strategy/parking/nearest_first_strategy.go
package parking
import "app/parking-lot/go-without-multithreading/entities"

type NearestFirstStrategy struct{}

func (s *NearestFirstStrategy) FindSpot(levels []*entities.Level, v entities.Vehicle) *entities.ParkingSpot {
	for _, level := range levels {
		if level.AvailableSpots[v.GetType()] > 0 {
			for _, spot := range level.Spots {
				if !spot.IsOccupied && spot.Type == v.GetType() {
					return spot
				}
			}
		}
	}
	return nil
}

File: parking-lot/java-without-multithreading/strategy/parking/NearestFirstStrategy.java
package strategy.parking;
import entities.Level;
import entities.ParkingSpot;
import entities.Vehicle;
import java.util.List;
import java.util.Optional;

public class NearestFirstStrategy implements ParkingStrategy {
    @Override
    public Optional<ParkingSpot> findSpot(List<Level> levels, Vehicle vehicle) {
        for (Level level : levels) {
            if (level.getAvailableSpots().get(vehicle.getType()) > 0) {
                for (ParkingSpot spot : level.getSpots()) {
                    if (!spot.isOccupied() && spot.getType() == vehicle.getType()) {
                        return Optional.of(spot);
                    }
                }
            }
        }
        return Optional.empty();
    }
}
